src/structlog/stdlib.py

```diff
@@ -407,15 +408,12 @@ def format(self, record):
         """
         Extract ``structlog``'s `event_dict` from ``record.msg`` and format it.
         """
+        should_format = False
         if isinstance(record.msg, dict):
-            # Both attached by wrap_for_formatter
-            logger = record._logger
-            meth_name = record._name
-
-            # We need to copy because it's possible that the same record gets
-            # processed by multiple logging formatters.  LogRecord.getMessage
-            # would transform our dict into a str.
-            ed = record.msg.copy()
+            logger, meth_name, ed = self.get_logger_name_and_msg(record)
+        elif self.is_json(record.msg):
```

Comment:
Perfomance-wise that must be really slow and conceptually, that does not look ideal. What if I decide to log a json-valid string ? I don't think there are any reason to treat this type of message differently. Moreover, if it's actually a json-string, you're parsing it twice. The proper way to do it would be a `try/except` in the else clause.

Not to appear closed-minded, but I reckon that the proper fix to this issue is once again something closer to what I suggested and started to implement in https://github.com/hynek/structlog/pull/115#issuecomment-300491382 . ProcesserHandler would wrap your Handler seamlessly. I do think that relying on the formatter to do this job wasn't the best idea in the first place. :-° But there's probably another, more elegant, more performant workaround.